### PR TITLE
Support for read from Cosmos DB Core SQL API

### LINKED ISSUE #23604: CosmosDB: Add support for Azure database CosmosDB with Core (SQL) Api
### What would you like to happen?

Add suppor for read data from Azure Cosmos DB with Core (SQL) API

### Issue Priority

Priority: 2

### Issue Component

Component: io-java-azure

### PR DESCRIPTION
**Pull request by RustedBones**
Fix #23604
Follow-up from #23610 

Implementation inspired from official [azure-cosmos-spark](https://github.com/Azure/azure-sdk-for-java/blob/5178502fdf053afb59cc71482e4972470afae730/sdk/cosmos/azure-cosmos-spark_3_2-12/README.md)

### CODE REVIEW

**Comment by RustedBones** [comment]
R: @kennknowles as you were part of the previous discussion in the linked PR

**Comment by RustedBones** [comment]
@Miuler can you give it a try in a real setup ? I've only tested with testcontainers.

**Comment by RustedBones** [comment]
@kennknowles & @Abacn can you have a look as you commented on https://github.com/apache/beam/pull/23610 ?

File: sdks/java/io/azure-cosmos/README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+# Cosmos DB Core SQL API
+
+Compile all module azure-cosmos
+
+```shell
+gradle sdks:java:io:azure-cosmos:build
+```
+
+Valite code:
```
[the remark this answers is not in the document]
Comment (RustedBones) [reply sdks/java/io/azure-cosmos/README.md:28]: Will remove the `README` as it is not informative. Most info here are common to the whole beam repo.

File: sdks/java/io/azure-cosmos/src/main/java/org/apache/beam/sdk/io/azure/cosmos/CosmosIO.java [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import reactor.core.publisher.Mono;
+
+@Experimental(Experimental.Kind.SOURCE_SINK)
+@SuppressWarnings({
+  "nullness" // TODO(https://github.com/apache/beam/issues/20497)
```
[the remark this answers is not in the document]
Comment (RustedBones) [reply sdks/java/io/azure-cosmos/src/main/java/org/apache/beam/sdk/io/azure/cosmos/CosmosIO.java:50]: I think I'll have to keep this one as the `CosmosAsyncClient` does not use nullable annotations.

File: sdks/java/io/azure-cosmos/src/main/java/org/apache/beam/sdk/io/azure/cosmos/CosmosIO.java [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        requestOptions.setMaxPrefetchPageCount(1);
+        requestOptions.setQuotaInfoEnabled(true);
+
+        estimatedByteSize =
+            container
+                .queryChangeFeed(requestOptions, ObjectNode.class)
+                .byPage()
+                .take(1)
+                .map(FeedResponse::getDocumentUsage)
+                .map(kb -> kb * 1000)
```
[the remark this answers is not in the document]
Comment (RustedBones) [reply sdks/java/io/azure-cosmos/src/main/java/org/apache/beam/sdk/io/azure/cosmos/CosmosIO.java:214]: right! Good catch. thanks!

File: sdks/java/io/azure-cosmos/src/main/java/org/apache/beam/sdk/io/azure/cosmos/CosmosOptions.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import org.apache.beam.sdk.options.Default;
+import org.apache.beam.sdk.options.DefaultValueFactory;
+import org.apache.beam.sdk.options.Description;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Strings;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+@Experimental(Experimental.Kind.SOURCE_SINK)
+@SuppressWarnings({
+  "nullness" // TODO(https://github.com/apache/beam/issues/20497)
```
[the remark this answers is not in the document]
Comment (RustedBones) [reply sdks/java/io/azure-cosmos/src/main/java/org/apache/beam/sdk/io/azure/cosmos/CosmosOptions.java:33]: A bit painful that the checker does not see that `!Strings.isNullOrEmpty()` removes null values. Will use plain check and remove anotation

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 5]
```diff
@@ -612,6 +612,7 @@ class BeamModulePlugin implements Plugin<Project> {
         aws_java_sdk2_regions                       : "software.amazon.awssdk:regions:$aws_java_sdk2_version",
         aws_java_sdk2_utils                         : "software.amazon.awssdk:utils:$aws_java_sdk2_version",
         aws_java_sdk2_profiles                      : "software.amazon.awssdk:profiles:$aws_java_sdk2_version",
+        azure_sdk_bom                               : "com.azure:azure-sdk-bom:1.2.14",
```
Comment (RustedBones) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:615]: I updated to use the `azure-sdk-bom` to ensure azure module compatibility

File: sdks/java/io/azure-cosmos/src/main/java/org/apache/beam/sdk/io/azure/cosmos/CosmosIO.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Abacn) [reply sdks/java/io/azure-cosmos/src/main/java/org/apache/beam/sdk/io/azure/cosmos/CosmosIO.java:56]: Thanks, then can we change the comment that currently links to #20497 to sth like "// CosmosAsyncClient does not use nullable annotations"

**Review by Abacn — APPROVED** [review]
Thanks! LGTM

**Pull request merged**